experimental/split_mlir/src/iree/split_mlir/OutlineFunctionsPassImpl.h

```diff
@@ -0,0 +1,285 @@
+                                   SmallVector<OpResult, 64>& rangeResults,
+                                   OpBuilder& builder) {
+  Region& region = *opRange.begin()->getParentRegion();
+  Block& dstBlock = region.emplaceBlock();
+  moveOpsIntoBlock(opRange, dstBlock);
+  replaceValueUsesWithNewBlockArguments(rangeOperands.begin(),
+                                        rangeOperands.end(), dstBlock);
+  addBlockReturn(dstBlock,
+                 ArrayRef<Value>(rangeResults.begin(), rangeResults.end()),
+                 builder);
+  func::FuncOp funcOp = builder.create<func::FuncOp>(
+      builder.getUnknownLoc(), name,
+      FunctionType::get(builder.getContext(), dstBlock.getArgumentTypes(),
+                        dstBlock.back().getOperandTypes()));
+  moduleOp.getBodyRegion().getBlocks().front().push_back(funcOp);
+  moveBlock(region, funcOp.getBody(), std::prev(region.end()),
+            funcOp.getBody().end());
+
+  return funcOp;
+}
+
+void createCall(func::FuncOp funcOp, Block& block, Block::iterator pos,
+                SmallVector<Value, 64>& rangeOperands,
+                SmallVector<OpResult, 64>& rangeResults, OpBuilder& builder) {
+  func::CallOp callOp = builder.create<func::CallOp>(
+      builder.getUnknownLoc(), funcOp,
+      ArrayRef<Value>(rangeOperands.begin(), rangeOperands.end()));
+  block.getOperations().insert(pos, callOp);
+  substititeUses(rangeResults.begin(), rangeResults.end(),
+                 callOp.getResults().begin(), callOp.getResults().end(),
+                 funcOp.getBody().back());
+}
+
+std::optional<func::FuncOp> outlineOpRange(
+    iterator_range<Block::iterator> opRange, StringRef name, ModuleOp moduleOp,
+    OpBuilder& builder) {
+  if (opRange.empty()) {
+    return std::nullopt;
+  }
+
+  SmallPtrSet<Operation*, 4> opsSet;
+  for (Operation& op : opRange) {
+    opsSet.insert(&op);
+  }
+  SmallVector<Value, 64> rangeOperands;
+  SmallVector<OpResult, 64> rangeResults;
+  std::tie(rangeOperands, rangeResults) =
+      getOperandsAndResultsForIsolation(opRange, opsSet);
+  Block& srcBlock = *opRange.begin()->getBlock();
+
+  func::FuncOp funcOp = createFunctionFromOps(
+      opRange, name, moduleOp, rangeOperands, rangeResults, builder);
+  createCall(funcOp, srcBlock, opRange.end(), rangeOperands, rangeResults,
+             builder);
+
+  return funcOp;
+}
+
+std::string getOutlinedFuncName(StringRef prefix, int blockIndex,
+                                int outlineRangeIndex) {
+  return (Twine(prefix) + "_outline_" + Twine(blockIndex) + "_" +
+          Twine(outlineRangeIndex))
+      .str();
+}
+
+void removeOutlineMarkers(iterator_range<Block::iterator> opRange) {
+  if (opRange.empty()) {
+    return;
+  }
+  opRange.begin()->removeAttr("outline_range_first");
+  std::prev(opRange.end())->removeAttr("outline_range_last");
+}
+
+LogicalResult outlineOpRanges(func::FuncOp funcOp, ModuleOp moduleOp,
```

Comment:
ditto, brief description explaining that each range is outlined and replaced with a function call within the function block, otherwise leaving any function related control flow alone.